# Comparing the neighborhoods of the two cities New York and Toronto) and determine how similar or dissimilar they are.

## Table of contents
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussion](#results)
* [Conclusion](#conclusion)

## Introduction: Business Problem <a name="introduction"></a>

### 1.1. Description:

New York and Toronto are two of the most well known and crowded places in two different progressive countries, US and Canada, respectively. Both of these cities are the major financial centers of the countries they are situated in. In this study I am going to use an API called Foursquare to explore the neighborhoods in the two cities. I will be using functions such as explore that come with Foursquare API to obtain most common venue categories in each neighborhoods of both the cities. Then build on this feature to arrange the neighborhoods into clusters. For clustering, I will be using k-means. Finally, I will be using Folium library to visualize the neighborhoods of both cities and see how similar or dissimilar they are. 

### 1.2. Audience

This problem targets people of New York City and Toronto. The analysis will help stakeholders identify similarities and dissimilarities between the people of both cities. Which can help stakeholders such a restaurant owners have a better clue if they really want to establish a branch in the other city or not. For example, if a person owns a restaurant in New York City and he wants to establish another branch in Toronto, this study will help him identify if such a venue as the one he is trying to establish will be a success or not.

### Data:

A data science problem mainly depends on the availability of data. Luckily, for our problem, we have both the datasets of the neighborhoods of New York City and Toronto available online. 

To be able to cluster and explore neighborhoods in those cities, we need datasets containing boroughs and neighborhoods of those boroughs. Other than that, we also need the coordinates (latitude, longitude) which we will be using with folium to visualize the neighborhoods properly on maps.

The neighborhood data for New York City is available on https://cocl.us/new_york_dataset in JSON format. We will be using our data analysis skills to analyze this JSON data and get only the neighborhoods data from it in the form of a Pandas dataframe. 

The neighborhood data for Toronto is available on Wikipedia https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M which we will read into a pandas dataframe by scrapping it from the Wikipedia page. Then we will clean the data by removing rows with missing “Borough” values.

## Importing all the required libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## We need data to work with. First we will download data for New York

In [2]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


Loading the data

In [3]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

Since we need only the data that we need to work with, which is the neighborhood data. We will clean and only get the neighborhood data

In [4]:
ny_neighborhoods_data = newyork_data['features']

We will convert the JSON data into a dataframe for analysis

In [5]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
ny_neighborhoods = pd.DataFrame(columns=column_names)

In [7]:
ny_neighborhoods.head()

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Now we will populate it with the JSON data

In [9]:
for data in ny_neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    ny_neighborhoods = ny_neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

This is how the data looks like

In [10]:
ny_neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

How many boroughs and neighborhoods?

In [11]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(ny_neighborhoods['Borough'].unique()),
        ny_neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


## Now that we have the neighborhood data for New York City, we need to have the neighborhood data for Toronto city in a dataframe for comparison.

We will need to scrape that data from Wikipedia

In [13]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
raw_data = pd.read_html(url)
raw_df = raw_data[0]

In [14]:
raw_df.head()

Postal code           Borough                Neighborhood
0         M1A      Not assigned                         NaN
1         M2A      Not assigned                         NaN
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

We need to get rid of the rows that have Borough written as "Not assigned"

In [15]:
raw_df = raw_df[raw_df.Borough != "Not assigned"]

In [16]:
raw_df.head()

Postal code           Borough                                  Neighborhood
2         M3A        North York                                     Parkwoods
3         M4A        North York                              Victoria Village
4         M5A  Downtown Toronto                    Regent Park / Harbourfront
5         M6A        North York             Lawrence Manor / Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

Now we have no missing values in our dataframe. 
But still we need latitude and longitude information for these neighborhoods

In [17]:
#Using the CSV option
csv_url = "http://cocl.us/Geospatial_data"
latlon_df = pd.read_csv(csv_url)
latlon_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Next we need to merge latitude longitude data with neighborhood data to make it one dataframe.

In [18]:
latlon_df.rename(columns={'Postal Code':'Postal code'},inplace=True)
can_df = pd.merge(raw_df, latlon_df, on='Postal code')
can_df.head()

Postal code           Borough                                  Neighborhood  \
0         M3A        North York                                     Parkwoods   
1         M4A        North York                              Victoria Village   
2         M5A  Downtown Toronto                    Regent Park / Harbourfront   
3         M6A        North York             Lawrence Manor / Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

Since we will be working with the neighborhood data of only those Boroughs that are of Toronto, we refining the dataframe even more

In [20]:
to_neighborhoods = can_df[can_df['Borough'].str.contains("Toronto")]

In [21]:
to_neighborhoods.reset_index(drop=True,inplace=True)

In [22]:
to_neighborhoods.head()

Postal code           Borough                                  Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park / Harbourfront   
1         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District / Ryerson   
3         M5C  Downtown Toronto                                St. James Town   
4         M4E      East Toronto                                   The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [23]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(to_neighborhoods['Borough'].unique()),
        to_neighborhoods.shape[0]
    )
)

The dataframe has 4 boroughs and 39 neighborhoods.


## Methodology <a name="methodology"></a>

Now we have the neighborhood data of the two cities (New York and Toronto) in two different dataframes "ny_neighborhoods" and "to_neighborhoods". We will use Foursquare API and compare the two cities to see how similar or dissimilar they are.

In this project I will be comparing two cities namely "New York" and "Toronto". I will show similarities and dissimilarities between the two cities, for example, what kind of venues, e.g. restaurants, are there in these two cities, do they have anything in common?

In this step, we will use Folium to create maps of New York and Toronto with their neighborhoods superimposed on top of them.

First we need latitude and logitude info for both cities:

In [25]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude_ny = location.latitude
longitude_ny = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


In [29]:
address1 = 'Toronto, ON'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address1)
latitude_to = location.latitude
longitude_to = location.longitude
print('The geograpical coordinate of Toront are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toront are 40.7127281, -74.0060152.


Map for New York City:

In [36]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude_ny, longitude_ny], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(ny_neighborhoods['Latitude'], ny_neighborhoods['Longitude'], ny_neighborhoods['Borough'], ny_neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

Map for Toronto City:

In [42]:
# create map of Manhattan using latitude and longitude values
map_toronto = folium.Map(location=[latitude_to, longitude_to], zoom_start=11)

# add markers to map
for lat, lng, label in zip(to_neighborhoods['Latitude'], to_neighborhoods['Longitude'], to_neighborhoods['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods of both cities and segment them.

We can see that the neighborhoods in New York are dense as compared to neighborhoods in Toronto.

#### Define Foursquare Credentials and Version

In [67]:
CLIENT_ID = '04O0JVTWOOFL3ZO4SMR3T3C3YUFARKYEGB5L0UT5DKCU3NAK' # your Foursquare ID
CLIENT_SECRET = 'VJDTRRG5DWMGXZ0ZOXMVAIZ53U3VE0JQVUTQYE2SZPJ3EDQ4' # your Foursquare Secret
VERSION = '20180605'

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 04O0JVTWOOFL3ZO4SMR3T3C3YUFARKYEGB5L0UT5DKCU3NAK
CLIENT_SECRET:VJDTRRG5DWMGXZ0ZOXMVAIZ53U3VE0JQVUTQYE2SZPJ3EDQ4


#### Let's explore the first neighborhood in both cities' dataframes

Get both the neighborhood's latitude and longitude values.

New York's first neighborhood:

In [68]:
neighborhood_latitude_ny = ny_neighborhoods.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude_ny = ny_neighborhoods.loc[0, 'Longitude'] # neighborhood longitude value

ny_neighborhood_name = ny_neighborhoods.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(ny_neighborhood_name, 
                                                               neighborhood_latitude_ny, 
                                                               neighborhood_longitude_ny))

Latitude and longitude values of Wakefield are 40.89470517661, -73.84720052054902.


Toronto's first neighborhood:

In [46]:
neighborhood_latitude_to = to_neighborhoods.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude_to = to_neighborhoods.loc[0, 'Longitude'] # neighborhood longitude value

to_neighborhood_name = to_neighborhoods.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(to_neighborhood_name, 
                                                               neighborhood_latitude_to, 
                                                               neighborhood_longitude_to))

Latitude and longitude values of Regent Park / Harbourfront are 43.6542599, -79.3606359.


#### Now, let's get the top 100 venues that are in Wakefield and Regent Park / Harbourfront within a radius of 500 meters.

First, let's create the GET request URL for both queries

For New York:

In [47]:
# type your answer here
LIMIT = 100
radius = 500
url_ny = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude_ny, 
    neighborhood_longitude_ny, 
    radius, 
    LIMIT)
url_ny # disp




'https://api.foursquare.com/v2/venues/explore?&client_id=04O0JVTWOOFL3ZO4SMR3T3C3YUFARKYEGB5L0UT5DKCU3NAK&client_secret=VJDTRRG5DWMGXZ0ZOXMVAIZ53U3VE0JQVUTQYE2SZPJ3EDQ4&v=20180605&ll=40.89470517661,-73.84720052054902&radius=500&limit=100'

For Toronto:

In [48]:
# type your answer here
LIMIT = 100
radius = 500
url_to = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude_to, 
    neighborhood_longitude_to, 
    radius, 
    LIMIT)
url_to # disp




'https://api.foursquare.com/v2/venues/explore?&client_id=04O0JVTWOOFL3ZO4SMR3T3C3YUFARKYEGB5L0UT5DKCU3NAK&client_secret=VJDTRRG5DWMGXZ0ZOXMVAIZ53U3VE0JQVUTQYE2SZPJ3EDQ4&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'

In [49]:
results_ny = requests.get(url_ny).json()
results_to = requests.get(url_to).json()

We will borrow the category extracting function from our week 3 assigment

In [55]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into  *pandas* dataframes.

In [56]:
venues_ny = results_ny['response']['groups'][0]['items']
    
nearby_venues_ny = json_normalize(venues_ny) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues_ny =nearby_venues_ny.loc[:, filtered_columns]

# filter the category for each row
nearby_venues_ny['venue.categories'] = nearby_venues_ny.apply(get_category_type, axis=1)

# clean columns
nearby_venues_ny.columns = [col.split(".")[-1] for col in nearby_venues_ny.columns]

nearby_venues_ny.head()

/home/saad/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name      categories        lat        lng
0  Lollipops Gelato    Dessert Shop  40.894123 -73.845892
1  Carvel Ice Cream  Ice Cream Shop  40.890487 -73.848568
2         Walgreens        Pharmacy  40.896528 -73.844700
3          Rite Aid        Pharmacy  40.896649 -73.844846
4           Dunkin'      Donut Shop  40.890459 -73.849089

The above dataframe contain venues around Wakefield, New York. Now we will do the same for Regent Park / Harbourfront, Toronto

In [57]:
venues_to = results_to['response']['groups'][0]['items']
    
nearby_venues_to = json_normalize(venues_to) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues_to =nearby_venues_to.loc[:, filtered_columns]

# filter the category for each row
nearby_venues_to['venue.categories'] = nearby_venues_to.apply(get_category_type, axis=1)

# clean columns
nearby_venues_to.columns = [col.split(".")[-1] for col in nearby_venues_to.columns]

nearby_venues_to.head()

/home/saad/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2      Morning Glory Cafe       Breakfast Spot  43.653947 -79.361149
3  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

In [58]:
nearby_venues_ny['categories'].value_counts()

Pharmacy          2
Gas Station       1
Sandwich Place    1
Laundromat        1
Dessert Shop      1
Ice Cream Shop    1
Donut Shop        1
Name: categories, dtype: int64

In [59]:
nearby_venues_to['categories'].value_counts()

Coffee Shop              7
Bakery                   3
Pub                      3
Park                     3
Theater                  2
Café                     2
Breakfast Spot           2
Event Space              1
Art Gallery              1
Spa                      1
Chocolate Shop           1
Historic Site            1
Health Food Store        1
Beer Store               1
Hotel                    1
Performing Arts Venue    1
Dessert Shop             1
Playground               1
Asian Restaurant         1
Mexican Restaurant       1
Yoga Studio              1
Gym / Fitness Center     1
Restaurant               1
Wine Shop                1
Ice Cream Shop           1
Farmers Market           1
Cosmetics Shop           1
French Restaurant        1
Antique Shop             1
Bank                     1
Electronics Store        1
Distribution Center      1
Shoe Store               1
Name: categories, dtype: int64

We can see that there are only a few categories of venues in Wakefield, New York, while in case of Regent Park / Harbourfront, Toronto, the categories are quite a few.We can also see that both the places have dessert shops and ice cream shops.

And how many venues were returned by Foursquare for Wakefield and Regent Park / Harbourfront?

In [60]:
print('{} venues were returned by Foursquare for Wakefield.'.format(nearby_venues_ny.shape[0]))

8 venues were returned by Foursquare for Wakefield.


In [61]:
print('{} venues were returned by Foursquare for Regent Park / Harbourfront.'.format(nearby_venues_to.shape[0]))

48 venues were returned by Foursquare for Regent Park / Harbourfront.


## Now Let's create repeat the same process to all the neighborhoods in New York and Toronto.

I will borrow the function that results the nearby venus from weeks assignment

In [75]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

First, lets get venues for all neighborhoods of New York:

In [78]:
newyork_venues = getNearbyVenues(names=ny_neighborhoods['Neighborhood'],
                                   latitudes=ny_neighborhoods['Latitude'],
                                   longitudes=ny_neighborhoods['Longitude']
                                  )

Wakefield
Co-op City
Eastchester
Fieldston
Riverdale
Kingsbridge
Marble Hill
Woodlawn
Norwood
Williamsbridge
Baychester
Pelham Parkway
City Island
Bedford Park
University Heights
Morris Heights
Fordham
East Tremont
West Farms
High  Bridge
Melrose
Mott Haven
Port Morris
Longwood
Hunts Point
Morrisania
Soundview
Clason Point
Throgs Neck
Country Club
Parkchester
Westchester Square
Van Nest
Morris Park
Belmont
Spuyten Duyvil
North Riverdale
Pelham Bay
Schuylerville
Edgewater Park
Castle Hill
Olinville
Pelham Gardens
Concourse
Unionport
Edenwald
Bay Ridge
Bensonhurst
Sunset Park
Greenpoint
Gravesend
Brighton Beach
Sheepshead Bay
Manhattan Terrace
Flatbush
Crown Heights
East Flatbush
Kensington
Windsor Terrace
Prospect Heights
Brownsville
Williamsburg
Bushwick
Bedford Stuyvesant
Brooklyn Heights
Cobble Hill
Carroll Gardens
Red Hook
Gowanus
Fort Greene
Park Slope
Cypress Hills
East New York
Starrett City
Canarsie
Flatlands
Mill Island
Manhattan Beach
Coney Island
Bath Beach
Borough Park
Dyker

In [153]:
newyork_venues.to_csv("newyork_venues.csv")

-





Now, lets get venues for all neighborhoods of Toronto:

In [79]:
toronto_venues = getNearbyVenues(names=to_neighborhoods['Neighborhood'],
                                   latitudes=to_neighborhoods['Latitude'],
                                   longitudes=to_neighborhoods['Longitude']
                                  )

Regent Park / Harbourfront
Queen's Park / Ontario Provincial Government
Garden District / Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond / Adelaide / King
Dufferin / Dovercourt Village
Harbourfront East / Union Station / Toronto Islands
Little Portugal / Trinity
The Danforth West / Riverdale
Toronto Dominion Centre / Design Exchange
Brockton / Parkdale Village / Exhibition Place
India Bazaar / The Beaches West
Commerce Court / Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park / The Junction South
North Toronto West
The Annex / North Midtown / Yorkville
Parkdale / Roncesvalles
Davisville
University of Toronto / Harbord
Runnymede / Swansea
Moore Park / Summerhill East
Kensington Market / Chinatown / Grange Park
Summerhill West / Rathnelly / South Hill / Forest Hill SE / Deer Park
CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst Quay / South Niagara / Island airport
Rosed

In [154]:
toronto_venues.to_csv("toronto_venues.csv")

-

#### Let's check the size of both the resulting dataframes

New York:

In [83]:
print(newyork_venues.shape)
newyork_venues.head()

(9754, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Wakefield              40.894705              -73.847201   
1    Wakefield              40.894705              -73.847201   
2    Wakefield              40.894705              -73.847201   
3    Wakefield              40.894705              -73.847201   
4    Wakefield              40.894705              -73.847201   

              Venue  Venue Latitude  Venue Longitude  Venue Category  
0  Lollipops Gelato       40.894123       -73.845892    Dessert Shop  
1  Carvel Ice Cream       40.890487       -73.848568  Ice Cream Shop  
2         Walgreens       40.896528       -73.844700        Pharmacy  
3          Rite Aid       40.896649       -73.844846        Pharmacy  
4           Dunkin'       40.890459       -73.849089      Donut Shop

Toronto:

In [84]:
print(toronto_venues.shape)
toronto_venues.head()

(1618, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park / Harbourfront               43.65426              -79.360636   
1  Regent Park / Harbourfront               43.65426              -79.360636   
2  Regent Park / Harbourfront               43.65426              -79.360636   
3  Regent Park / Harbourfront               43.65426              -79.360636   
4  Regent Park / Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

We can see that the number of venues in New York (9754) is far greater than the number of venues in Toronto (1618)

#### Let's find out how many unique categories can be curated from all the returned venues

Unique catergories in New York:

In [86]:
print('There are {} uniques categories in New York.'.format(len(newyork_venues['Venue Category'].unique())))

There are 426 uniques categories in New York.


Unique catergories in Toronto:

In [87]:
print('There are {} uniques categories in Toronto.'.format(len(toronto_venues['Venue Category'].unique())))

There are 232 uniques categories in Toronto.


Also, New York has more unique categories than Toronto as it is the worlds most busy place.

## Now lets analyze each neighborhood in both cities.

### New York:

In [88]:
# one hot encoding
newyork_onehot = pd.get_dummies(newyork_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
newyork_onehot['Neighborhood'] = newyork_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [newyork_onehot.columns[-1]] + list(newyork_onehot.columns[:-1])
newyork_onehot = newyork_onehot[fixed_columns]
newyork_onehot
newyork_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  \
0            0                  0               0                  0   
1            0                  0               0                  0   
2            0                  0               0                  0   
3            0                  0               0                  0   
4            0                  0               0                  0   

   African Restaurant  American Restaurant  Animal Shelter  Antique Shop  \
0                   0                    0               0             0   
1                   0                    0               0             0   
2                   0                    0               0             0   
3                   0                    0               0             0   
4                   0                    0               0             0   

   Arcade  Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0       0                 0                       0            0           0   
1       0                 0                       0            0           0   
2       0                 0                       0            0           0   
3       0                 0                       0            0           0   
4       0                 0                       0            0           0   

   Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0                    0                     0                 0   
1                    0                     0                 0   
2                    0                     0                 0   
3                    0                     0                 0   
4                    0                     0                 0   

   Athletics & Sports  Auditorium  Australian Restaurant  Austrian Restaurant  \
0                   0           0                      0                    0   
1                   0           0                      0                    0   
2                   0           0                      0                    0   
3                   0           0                      0                    0   
4                   0           0                      0                    0   

   Auto Garage  Auto Workshop  Automotive Shop  BBQ Joint  Baby Store  \
0            0              0                0          0           0   
1            0              0                0          0           0   
2            0              0                0          0           0   
3            0              0                0          0           0   
4            0              0                0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 0   
3           0       0     0    0               0                 0   
4           0       0     0    0               0                 0   

   Basketball Court  Basketball Stadium  Bath House  Beach  Beach Bar  \
0                 0                   0           0      0          0   
1                 0                   0           0      0          0   
2                 0                   0           0      0          0   
3                 0                   0           0      0          0   
4                 0                   0           0      0          0   

   Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0                0         0            0           0              0   
1                0         0            0           0              0   
2                0         0            0           0              0   
3                0         0            0           0              0   
4                0         0            0           0              0   

   Bike Rental /

In [89]:
newyork_onehot.shape

(9754, 426)

### Toronto:

In [90]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot
toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       1     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               1        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Butcher  Café  Cajun / Creole Restaurant  \
0              0         0        0     0                          0   
1              0         0        0     0                          0   
2              0         0        0     0                          0   
3              0         0        0     0                          0   
4              0         0        0     0                          0   

   C

In [92]:
toronto_onehot.shape

(1618, 232)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

### New York

In [136]:
newyork_grouped = newyork_onehot.groupby('Neighborhood').mean().reset_index()
newyork_grouped.head()

Neighborhood  Yoga Studio  Accessories Store  Adult Boutique  \
0       Allerton          0.0                0.0             0.0   
1       Annadale          0.0                0.0             0.0   
2  Arden Heights          0.0                0.0             0.0   
3      Arlington          0.0                0.0             0.0   
4       Arrochar          0.0                0.0             0.0   

   Afghan Restaurant  African Restaurant  American Restaurant  Animal Shelter  \
0                0.0                 0.0             0.000000             0.0   
1                0.0                 0.0             0.000000             0.0   
2                0.0                 0.0             0.000000             0.0   
3                0.0                 0.0             0.142857             0.0   
4                0.0                 0.0             0.000000             0.0   

   Antique Shop  Arcade  Arepa Restaurant  Argentinian Restaurant  \
0           0.0     0.0               0.0                     0.0   
1           0.0     0.0               0.0                     0.0   
2           0.0     0.0               0.0                     0.0   
3           0.0     0.0               0.0                     0.0   
4           0.0     0.0               0.0                     0.0   

   Art Gallery  Art Museum  Arts & Crafts Store  Arts & Entertainment  \
0          0.0         0.0                  0.0                   0.0   
1          0.0         0.0                  0.0                   0.0   
2          0.0         0.0                  0.0                   0.0   
3          0.0         0.0                  0.0                   0.0   
4          0.0         0.0                  0.0                   0.0   

   Asian Restaurant  Athletics & Sports  Auditorium  Australian Restaurant  \
0               0.0            0.000000         0.0                    0.0   
1               0.0            0.000000         0.0                    0.0   
2               0.0            0.000000         0.0                    0.0   
3               0.0            0.000000         0.0                    0.0   
4               0.0            0.047619         0.0                    0.0   

   Austrian Restaurant  Auto Garage  Auto Workshop  Automotive Shop  \
0                  0.0          0.0            0.0              0.0   
1                  0.0          0.0            0.0              0.0   
2                  0.0          0.0            0.0              0.0   
3                  0.0          0.0            0.0              0.0   
4                  0.0          0.0            0.0              0.0   

   BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank  Bar  Baseball Field  \
0        0.0         0.0    0.000000  0.032258   0.0  0.0             0.0   
1        0.0         0.0    0.111111  0.000000   0.0  0.0             0.0   
2        0.0         0.0    0.000000  0.000000   0.0  0.0             0.0   
3        0.0         0.0    0.000000  0.000000   0.0  0.0             0.0   
4        0.0         0.0    0.047619  0.000000   0.0  0.0             0.0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Bath House  Beach  \
0               0.0               0.0                 0.0         0.0    0.0   
1               0.0               0.0                 0.0         0.0    0.0   
2               0.0               0.0                 0.0         0.0    0.0   
3               0.0               0.0                 0.0         0.0    0.0   
4               0.0               0.0                 0.0         0.0    0.0   

   Beach Bar  Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  \
0        0.0              0.0       0.0          0.0         0.0   
1        0.0              0.0       0.0          0.0         0.0   
2        0.0              0.0       0.0          0.0         0.0   
3        0.0              0.0       0.0          0.0         0.0   
4        0.0              0.0       0.0          0.0         0.0   

   Bi

In [137]:
newyork_grouped.shape

(299, 426)

### Toronto

In [95]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio   Airport  \
0                                        Berczy Park     0.000000  0.000000   
1     Brockton / Parkdale Village / Exhibition Place     0.000000  0.000000   
2              Business reply mail Processing Centre     0.000000  0.000000   
3  CN Tower / King and Spadina / Railway Lands / ...     0.000000  0.066667   
4                                 Central Bay Street     0.015873  0.000000   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0            0.000000      0.000000        0.000000         0.000000   
1            0.000000      0.000000        0.000000         0.000000   
2            0.000000      0.000000        0.000000         0.000000   
3            0.066667      0.066667        0.133333         0.133333   
4            0.000000      0.000000        0.000000         0.000000   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0          0.000000                  0.0           0.0       0.0     0.017241   
1          0.000000                  0.0           0.0       0.0     0.000000   
2          0.000000                  0.0           0.0       0.0     0.000000   
3          0.066667                  0.0           0.0       0.0     0.000000   
4          0.000000                  0.0           0.0       0.0     0.000000   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0    0.000000                  0.0               0.0                 0.0   
1    0.000000                  0.0               0.0                 0.0   
2    0.000000                  0.0               0.0                 0.0   
3    0.000000                  0.0               0.0                 0.0   
4    0.015873                  0.0               0.0                 0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank       Bar  \
0       0.000000   0.017241         0.0    0.017241  0.034483   0.0  0.000000   
1       0.000000   0.000000         0.0    0.000000  0.045455   0.0  0.045455   
2       0.066667   0.000000         0.0    0.000000  0.000000   0.0  0.000000   
3       0.000000   0.000000         0.0    0.000000  0.000000   0.0  0.000000   
4       0.000000   0.000000         0.0    0.000000  0.000000   0.0  0.031746   

   Baseball Stadium  Basketball Stadium     Beach  Bed & Breakfast  Beer Bar  \
0               0.0            0.017241  0.017241              0.0  0.034483   
1               0.0            0.000000  0.000000              0.0  0.000000   
2               0.0            0.000000  0.000000              0.0  0.000000   
3               0.0            0.000000  0.000000              0.0  0.000000   
4               0.0            0.000000  0.000000              0.0  0.000000   

   Beer Store  Belgian Restaurant    Bistro  Board Shop  Boat or Ferry  \
0         0.0                 0.0  0.017241         0.0       0.000000   
1         0.0                 0.0  0.000000         0.0       0.000000   
2         0.0                 0.0  0.000000         0.0       0.000000   
3         0.0                 0.0  0.000000         0.0       0.066667   
4         0.0                 0.0  0.000000         0.0       0.000000   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot   Brewery  \
0        0.0  0.000000                   0.0        0.017241  0.000000   
1        0.0  0.000000                   0.0        0.090909  0.000000   
2        0.0  0.000000                   0.0        0.000000  0.066667   
3        0.0  0.066667                   0.0        0.000000  0.000000   
4        0.0  0.000000                   0.0        0.000000  0.000000   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line   Butcher  \
0         0.000000       0.0      0.000000       0.000000       0.0  0.017241   
1         0.000000       0.0      0.000000       0.045455       0.0  0.000000   
2         0.000000       0.0      0.000000       0.066667       0.0  0.000000   
3     

In [96]:
toronto_grouped.shape

(39, 232)

#### Now Let's print each neighborhood of both cities along with the top 5 most common venues

### New York

In [ ]:
num_top_venues = 5

for hood in newyork_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = newyork_grouped[newyork_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

Now putting it into a dataframe

First, let's write a function to sort the venues in descending order.

In [121]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [138]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted_ny = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_ny['Neighborhood'] = newyork_grouped['Neighborhood']

for ind in np.arange(newyork_grouped.shape[0]):
    neighborhoods_venues_sorted_ny.iloc[ind, 1:] = return_most_common_venues(newyork_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_ny.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0       Allerton           Pizza Place           Supermarket   
1       Annadale      Sushi Restaurant            Bagel Shop   
2  Arden Heights           Pizza Place         Deli / Bodega   
3      Arlington              Bus Stop          Home Service   
4       Arrochar    Italian Restaurant         Deli / Bodega   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Deli / Bodega      Department Store          Intersection   
1                 Diner           Pizza Place            Restaurant   
2              Pharmacy           Coffee Shop              Bus Stop   
3         Deli / Bodega   American Restaurant         Boat or Ferry   
4              Bus Stop            Taco Place           Supermarket   

  6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
0     Martial Arts Dojo  Check Cashing Service    Chinese Restaurant   
1         Train Station                   Food            Sports Bar   
2          Home Service   Fast Food Restaurant  Ethiopian Restaurant   
3           Coffee Shop          Women's Store         Event Service   
4   Sporting Goods Shop         Sandwich Place     Polish Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0                Bakery      Electronics Store  
1        Cosmetics Shop   Ethiopian Restaurant  
2         Event Service            Event Space  
3           Event Space                Exhibit  
4           Pizza Place     Athletics & Sports

### Toronto

In [ ]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

Now putting it into a dataframe

In [101]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted_to = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_to['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted_to.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_to.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1     Brockton / Parkdale Village / Exhibition Place                  Café   
2              Business reply mail Processing Centre    Light Rail Station   
3  CN Tower / King and Spadina / Railway Lands / ...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar           Cheese Shop                  Café   
1        Breakfast Spot           Coffee Shop          Climbing Gym   
2                Garden            Restaurant         Auto Workshop   
3       Airport Service           Coffee Shop       Harbor / Marina   
4                  Café    Italian Restaurant        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0                Bakery    Seafood Restaurant                 Restaurant   
1         Burrito Place            Restaurant         Italian Restaurant   
2  Fast Food Restaurant        Farmers Market                        Spa   
3                 Plane   Rental Car Location           Sculpture Garden   
4                   Bar        Ice Cream Shop  Middle Eastern Restaurant   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0              Beer Bar  Portuguese Restaurant                Butcher  
1               Stadium           Intersection                    Bar  
2               Brewery             Smoke Shop          Garden Center  
3              Boutique          Boat or Ferry       Airport Terminal  
4          Burger Joint        Thai Restaurant            Salad Place

# Cluster Neighborhoods

## First let's do clustering for New York

Run *k*-means to cluster the neighborhood into 10 clusters.

In [139]:
# set number of clusters
kclusters = 10
newyork_grouped_clustering = newyork_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(newyork_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([6, 6, 6, 7, 9, 3, 3, 6, 3, 3], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [140]:
# add clustering labels
neighborhoods_venues_sorted_ny.insert(0, 'Cluster Labels', kmeans.labels_)
newyork_merged = ny_neighborhoods

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
newyork_merged = newyork_merged.join(neighborhoods_venues_sorted_ny.set_index('Neighborhood'), on='Neighborhood')

newyork_merged.head() # check the last columns!

Borough Neighborhood   Latitude  Longitude  Cluster Labels  \
0   Bronx    Wakefield  40.894705 -73.847201             8.0   
1   Bronx   Co-op City  40.874294 -73.829939             8.0   
2   Bronx  Eastchester  40.887556 -73.827806             8.0   
3   Bronx    Fieldston  40.895437 -73.905643             1.0   
4   Bronx    Riverdale  40.890834 -73.912585             1.0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0              Pharmacy            Laundromat        Ice Cream Shop   
1           Bus Station    Chinese Restaurant           Pizza Place   
2  Caribbean Restaurant                 Diner           Bus Station   
3           Bus Station                 Plaza         Women's Store   
4           Bus Station                  Park            Food Truck   

  4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
0          Dessert Shop             Donut Shop           Gas Station   
1        Baseball Field         Mattress Store         Grocery Store   
2         Deli / Bodega               Platform            Donut Shop   
3  Fast Food Restaurant  Entertainment Service  Ethiopian Restaurant   
4                   Gym           Home Service            Playground   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0        Sandwich Place         Women's Store  Ethiopian Restaurant   
1        Ice Cream Shop              Pharmacy            Bagel Shop   
2          Intersection    Seafood Restaurant           Pizza Place   
3         Event Service           Event Space               Exhibit   
4                 Plaza                  Bank           Film Studio   

  10th Most Common Venue  
0          Event Service  
1                   Park  
2          Bowling Alley  
3             Eye Doctor  
4                Factory

In [141]:
newyork_merged['Cluster Labels'].isnull().sum()

3

In [142]:
newyork_merged['Cluster Labels'] = newyork_merged['Cluster Labels'].fillna(0)

In [144]:
newyork_merged['Cluster Labels'] = newyork_merged['Cluster Labels'].astype(int)

Finally, let's visualize the resulting clusters

In [146]:
# create map
map_clusters_ny = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(newyork_merged['Latitude'], newyork_merged['Longitude'], newyork_merged['Neighborhood'], newyork_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters_ny)
       
map_clusters_ny

In [111]:
newyork_merged.rename(columns={'Cluster Labels':'Labels'},inplace=True)

## Now let's do clustering for Toronto

Run *k*-means to cluster the neighborhood into 10 clusters.

In [147]:
# set number of clusters
kclusters = 10
toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 3, 0, 0, 3, 7, 3, 3, 3, 5], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [148]:
# add clustering labels
neighborhoods_venues_sorted_to.insert(0, 'Cluster Labels', kmeans.labels_)
toronto_merged = to_neighborhoods

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted_to.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal code           Borough                                  Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park / Harbourfront   
1         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District / Ryerson   
3         M5C  Downtown Toronto                                St. James Town   
4         M4E      East Toronto                                   The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               3        Clothing Store   
3  43.651494 -79.375418               3           Coffee Shop   
4  43.676357 -79.293031               9     Health Food Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Pub                Bakery                  Park   
1      Sushi Restaurant                 Diner           Yoga Studio   
2           Coffee Shop                  Café        Cosmetics Shop   
3                  Café          Cocktail Bar             Gastropub   
4                 Trail                   Pub         Women's Store   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0        Breakfast Spot                  Café                    Theater   
1              Beer Bar        Sandwich Place              Burrito Place   
2       Bubble Tea Shop            Restaurant  Middle Eastern Restaurant   
3   American Restaurant        Clothing Store             Cosmetics Shop   
4         Deli / Bodega  Ethiopian Restaurant          Electronics Store   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Yoga Studio    Mexican Restaurant             Shoe Store  
1                         Café                  Park               Creperie  
2          Japanese Restaurant    Italian Restaurant                Theater  
3                     Creperie        Lingerie Store       Department Store  
4  Eastern European Restaurant            Donut Shop       Doner Restaurant

In [149]:
toronto_merged['Cluster Labels'].isnull().sum()

0

In [142]:
newyork_merged['Cluster Labels'] = newyork_merged['Cluster Labels'].fillna(0)

In [144]:
newyork_merged['Cluster Labels'] = newyork_merged['Cluster Labels'].astype(int)

Finally, let's visualize the resulting clusters

In [152]:
# create map
map_clusters_to = folium.Map(location=[latitude_to, longitude_to], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters_to)
       
map_clusters_to

## Results and Discussion <a name="results"></a>

Our analysis shows that New York is far denser than Toronto neighborhood-wise. We also saw that there are only a few categories of venues in Wakefield, New York, while in case of Regent Park / Harbourfront, Toronto, the categories are quite a few.Two venues, namely, dessert shops and ice cream shops were common to both Wakefield and  Regent Park / Harbourfront. Only 8 venues were returned by Foursqaure API for wakefield, while 48 venues were returned by the API for Rogent Park / Harbourfront. The total number of venues for all of the neighborhoods in New York is 9754 while the number of venues for each neighborhood in Toronto is 1618. We also found out that there were 426 unique categories of venues in New York and 232 unique categories in case of Toronto. Then we made dataframes of the top most common venues for each neighborhood in both the cities. And finally, we clustered the neighborhoods in both cities using k-means clustering and showed the clusters on maps.

## Conclusion <a name="conclusion"></a>

The purpose of this project was to explore neighborhoods and venues near those neighborhoods in two major cities of the world, namely, Toronto and New York. Second most important purpose of the project was to put to practice what we learned about Data Analysis and Foursquare API to real world problems and test our understanding of those skills. 